proofread: performance_evaluation_guide.md (#1553)
juicedata/juicefs: docs/en/benchmark/performance_evaluation_guide.md

---
# JuiceFS Performance Evaluation Guide

Before starting performance testing, it is a good idea to write down a general description of usage scenario, including:

1. What is the application for? For example, Apache Spark, PyTorch, or a program you developed yourself

> 2. The requisite resource for running the application, including CPU, memory, network, and node size

Reviewer: 加上“allocation”是不是更符合原文“资源配置”的意思呢？:)

3. The estimated data size, including the number of files and their volume
4. The file size and access mode (large or small files, sequential or random reads and writes)

> 5. Performance requirements, such as the amount of data to be written or read per second, QPS, or the latency for the operation, etc

Reviewer: ```suggestion
5. Performance requirements, such as the amount of data to be written or read per second, QPS, operation latency, etc.
```

### Attention

JuiceFS v1.0+ has Trash enabled by default, which means the benchmark tools will create and delete temporary files in the file system. These files will eventually be dumped to the `.trash` folder which consumes storage space. To avoid this, you can disable the Trash before benchmarking by running `juicefs config META-URL --trash-days 0`. See [trash](../security/trash.md) for details.

### JuiceFS Bench

> The command `bench` of JuiceFS can help you do a quick performance test on a standalone machine. With the test results, you can see if you have your environment correctly configured, and evaluate if JuiceFS works at a good performance in your environment. Assuming you have mounted JuiceFS to `/mnt/jfs` on your server (if you need help with initializing or mounting JuiceFS, please refer to the [Quick Start Guide](../getting-started/for_local.md)), the following command will do this test (the `-p` parameter is recommended to set to the number of CPU cores on the server).

Reviewer: ```suggestion
The JuiceFS `bench` command can help you do a quick performance test on a standalone machine. With the test results, it is easy to evaluate if your environment configuration and JuiceFS performance are normal. Assuming you have mounted JuiceFS to `/mnt/jfs` on your server, execute the following command for this test (the `-p` parameter is recommended to set to the number of CPU cores on the server). If you need help with initializing or mounting JuiceFS, please refer to the [Quick Start Guide](../getting-started/for_local.md))
```
1. 以前的版本似乎更简洁书面化一些
2. 我把括号里的内容移到了最后，因为括号会让人在读的时候停顿一下，造成上半句和下半句“断截”

```bash
juicefs bench /mnt/jfs -p 4
```

The test results will show each performance indicator in green, yellow or red. If you see red indicators in your results, please check the relevant configuration first. Feel free to post any problems you encountered in detail on [GitHub Discussions](https://github.com/juicedata/juicefs/discussions).

![bench](../images/bench-guide-bench.png)

> The JuiceFS `bench` performance test flows as follows (Its a simple logic. Please take a look at the [source code](https://github.com/juicedata/juicefs/blob/main/cmd/bench.go) if you want more details).

Reviewer: ```suggestion
The detailed JuiceFS `bench` performance test flows are shown below (The logic behind is very simple. Please take a look at the [source code](https://github.com/juicedata/juicefs/blob/main/cmd/bench.go) if you are interested).
```

> 1. N concurrent `write`, each to a large file of 1 GiB with IO size 1 MiB

Reviewer: ```suggestion
1. N concurrent `write`, each to a large file of 1 GiB with IO size of 1 MiB
```

> 2. N concurrent `read`, each from a large file of 1 GiB previously written, with IO size 1 MiB

Reviewer: ```suggestion
2. N concurrent `read`, each from the large file of 1 GiB previously written, with IO size of 1 MiB
```

> 3. N concurrent `write`, each to 100 small files of 128 KiB, with IO size 128 KiB

Reviewer: ```suggestion
3. N concurrent `write`, each to 100 small files of 128 KiB, with IO size of 128 KiB
```

4. N concurrent `read`, each from 100 small files of 128 KiB previously written, with IO size 128 KiB

> 5. N concurrent `stat`, each on 100 128 KiB small files previously written

Reviewer: ```suggestion
5. N concurrent `stat`, each on the 100 small files of 128 KiB previously written 
```

6. clean up the temporary directory for testing

> The concurrency scale N could be provided through the `-p` parameter of the `bench` command.

Reviewer: 原来的版本好像更符合中文的意思？

Here's a performance comparison using a few common storage types provided by AWS.

> - EFS with capacity 1TiB performs 150MiB/s of `read` and 50MiB/s of `write`, which costs $0.08/GB-month.

Reviewer: ```suggestion
- EFS with 1TiB capacity performs 150MiB/s of `read` and 50MiB/s of `write` at a cost of $0.08/GB-month.
```

- EBS st1 is a throughput-optimized HDD with a maximum throughput of 500MiB/s, a maximum IOPS (1MiB I/O) of 500, and a maximum capacity of 16TiB, priced at $0.045/GB-month.

> - EBS gp2 is a general-purpose SSD with a maximum throughput of 250MiB/s, maximum IOPS (16KiB I/O) of 16,000, and maximum capacity of 16TiB, which costs $0.10/GB-month.

Reviewer: ```suggestion
- EBS gp2 is a universal SSD with a maximum throughput of 250MiB/s, maximum IOPS (16KiB I/O) of 16,000, and a maximum capacity of 16TiB, priced at $0.10/GB-month.
```

It is easy to see that, in the above tests, JuiceFS has significantly better sequential read and write capabilities than AWS EFS and more throughput than the commonly used EBS. However, writing small files is not as fast because every file written needs to be persisted to S3 and there is typically a fixed overhead of 10-30ms on calling the object storage API.

:::note
The performance of Amazon EFS is linearly related to capacity ([refer to the official documentation](https://docs.aws.amazon.com/efs/latest/ug/performance.html#performancemodes)), which makes it unsuitable for being used in high throughput scenarios with small data sizes.
:::

:::note

> Prices come from [AWS US East, Ohio Region](https://aws.amazon.com/ebs/pricing/?nc1=h_ls), differing slightly among regions.

Reviewer: ```suggestion
Prices refer to [AWS US East, Ohio Region](https://aws.amazon.com/ebs/pricing/?nc1=h_ls), differing slightly among regions.
```

### JuiceFS Stats

> JuiceFS `stats` is a tool for real-time statistics of JuiceFS performance metrics, similar to the `dstat` command on Linux systems, which displays changes of metrics for JuiceFS clients in real time (see [documentation](stats_watcher.md) for detailed). When the command `juicefs bench` is running, create a new session and execute the following command.

Reviewer: ```suggestion
JuiceFS `stats` is a tool for real-time statistics of JuiceFS performance metrics, similar to the `dstat` command on Linux systems. It can display changes of metrics for JuiceFS clients in real-time (see [documentation](stats_watcher.md) for details). For this,  create a new session and execute the following command when the command `juicefs bench` is running, 
```

```bash
juicefs stats /mnt/jfs --verbosity 1
```

> The results are as follows, which can be more easy to understand when viewing against the benchmarking process described above.

Reviewer: ```suggestion
The results are shown below, which would be easier to understand when combing with the `bench` performance test flows described above.
```

尝试着benchmarking process改成了与上文的“bench performance test flows” (基准测试流程）一致

![stats](../images/bench-guide-stats.png)

The meaning of indicators is as follows:

- usage

>   - cpu: the CPU usage of JuiceFS process

Reviewer: ```suggestion
  - cpu: the CPU consumed by JuiceFS process
```

>   - mem: the physical memory usage of JuiceFS process

Reviewer: ```suggestion
  - mem: the physical memory consumed by JuiceFS process
```

  - buf: internal read/write buffer size of JuiceFS process, limited by mount option `--buffer-size`
  - cache: internal metric, can be simply ignored
- fuse

>   - ops/lat: requests per second processed by the FUSE interface and their average latency (in milliseconds)

Reviewer: 原来的表述会不会更准确一点？强调了the number of requests

  - read/write: bandwidth of the FUSE interface to handle read and write requests per second
- meta

>   - ops/lat: requests per second processed by the metadata engine and their average latency (in milliseconds). Please note that some requests that can be processed directly in the cache are not included in the statistics, in order to better reflect the time spent by the client interacting with the metadata engine.

Reviewer: ```suggestion
  - ops/lat: the number of requests per second processed by the metadata engine and their average latency (in milliseconds). Please note that some requests that can be directly processed in the cache are not included in the statistics, in order to better reflect the time spent by the client interacting with the metadata engine.
```

>   - txn/lat: **write transactions** per second processed by the metadata engine and their average latency (in milliseconds). Read-only requests such as `getattr` are only counted as ops but not txn.

Reviewer: The same comment as above. :P and a few suggestions

```suggestion
  - txn/lat: the number of **write transactions** processed by the metadata engine per second and their average latency (in milliseconds). Read-only requests such as `getattr` are only counted as ops but not txn.
```

>   - retry: **write transactions** per second that the metadata engine retries

Reviewer: ```suggestion
  - retry: the number of **write transactions** that the metadata engine retries per second 
```

- blockcache
  - read/write: read/write traffic per second for the local data cache of the client
- object

>   - get/get_c/lat: bandwidth, requests per second, and their average latency (in milliseconds) for object storage processing **read requests**

Reviewer: ```suggestion
  - get/get_c/lat: bandwidth, number of requests and average latency (in milliseconds) of object storage processing **read requests** per second
```
我理解的意思是 get/get_c/lat = 对象存储没秒处理读请求时的带宽值，请求数量和平均延时

>   - put/put_c/lat: bandwidth, requests per second, and their average latency (in milliseconds) for object storage processing **write requests**

Reviewer: 按照相同的理解，我把这句也改了下

```suggestion
  - put/put_c/lat: bandwidth, number of requests, and their average latency (in milliseconds) of object storage processing **write requests** per second
```

>   - del_c/lat: **delete requests** per second the object storage can process, and the average latency (in milliseconds)

Reviewer: ```suggestion
  - del_c/lat: the number of **delete requests** that object storage can process per second, and the average latency (in milliseconds)
```

### JuiceFS Profile

> JuiceFS `profile` is used to output all access logs of the JuiceFS client in real time, including information about each request. It can also be used to play back and count the JuiceFS access logs, allowing users to visualize the running status of JuiceFS (for detailed description and usage. See [documentation](operations_profiling.md)). Run the following command in another session while the `juicefs bench` command is running.

Reviewer: ```suggestion
JuiceFS `profile` is used to output all access logs of the JuiceFS client in real time, including information about each request. It can also be used to play back and count JuiceFS access logs, and visualize the JuiceFS running status (see [documentation](operations_profiling.md) for details). To run the JuiceFS profile, execute the following command in another session while the `juicefs bench` command is running.
```

```bash
cat /mnt/jfs/.accesslog > access.log
```

`.accessslog` is a virtual file for JuiceFS access logs. It does not produce any data until it is read (e.g. by executing `cat`). Press <kbd>Ctrl</kbd>-<kbd>C</kbd> to terminate the `cat` command and run the following one.

```bash
juicefs profile access.log --interval 0
```

> The `---interval` parameter sets the sampling interval for accessing the log. 0 means quickly replay the log file to generate statistics, as shown in the following figure.

Reviewer: ```suggestion
The `---interval` parameter sets the sampling interval for accessing the log. When it is set to 0,  a specified log file can be quickly replayed to generate statistics, as shown in the following figure.
```

![profile](../images/bench-guide-profile.png)

> From the description of the previous benchmarking process, a total of (1 + 100) * 4 = 404 files were created during this test, and each file went through the process of "Create → Write → Close → Open → Read → Close → Delete". So there are a total of:

Reviewer: ```suggestion
Based on the bench performance test flows as described above, a total of (1 + 100) * 4 = 404 files were created during this test, and each file went through the process of "Create → Write → Close → Open → Read → Close → Delete". So there are a total of:
```

- `--name`: user-specified test name, which affects the test file name
- `--directory`: test directory
- `--ioengine`: the way to send IO when testing; usually `libaio` is used

> - `--rw`: common options are read, write, randread and randwrite, which stand for sequential read/write and random read/write respectively

Reviewer: ```suggestion
- `--rw`: commonly used are read, write, randread and randwrite, which stand for sequential read/write and random read/write, respectively
```

common option 在这个语境里好像会有歧义

- `--bs`: the size of each IO
- `--size`: the total size of IO per thread; usually equal to the size of the test file

> - `--numjobs`: number of concurrent test threads; each thread runs with an individual test file by default

Reviewer: ```suggestion
- `--numjobs`: number of concurrent test threads; each thread runs an individual test file by default
```

> - `--direct`: add the `O_DIRECT` flag bit on opening a file to disable system buffering, which can make the test results more stable and accurate

Reviewer: ```suggestion
- `--direct`: add the `O_DIRECT` flag bit when opening a file without using system buffering, which can make the test results more stable and accurate
```
原来的翻译好像更符合中文版本的意思？

READ: bw=93.6MiB/s (98.1MB/s), 93.6MiB/s-93.6MiB/s (98.1MB/s-98.1MB/s), io=4096MiB (4295MB), run=43773-43773msec
```

### Vdbench Multi-machine Performance Test

> Vdbench is a commonly used file system evaluation tool, and supports multi-machine concurrent testing well.

Reviewer: ```suggestion
Vdbench is a commonly used file system evaluation tool, and it supports concurrent multi-machine testing very well.
```
我觉得原来的翻译挺好的 :P 
我觉得好像cocurrent multi-machine testing 更准确一些？但是我也不是很懂

#### Working Environment

> Similar to the JuiceFS Bench test environment, but two more (three in total) hosts with same hardware specifications.

Reviewer: ```suggestion
Similar to the JuiceFS Bench test environment, but with two more hosts (three in total) with the same hardware specifications.
```

#### Preparation

vdbench needs to be installed under the same path on each node:

1. Download version 50406 from the [Official Website](https://www.oracle.com/downloads/server-storage/vdbench-downloads.html)
2. Install Java: `apt-get install openjdk-8-jre`
3. Verify that vdbench is installed successfully: `./vdbench -t`

> We assume the names of the three nodes are `node0`, `node1` and `node2`. You need to create a configuration file on `node0` as follows (to test reading and writing a large number of small files):

Reviewer: ```suggestion
Assuming the names of the three nodes are `node0`, `node1` and `node2`, then you need to create a configuration file on `node0` as follows (to test reading and writing a large number of small files):
```

READ_OPENS          Files opened for read activity:             188,317        627/sec
```

> The overall rate of 128 KiB file creating is 498 (files/s), while file reading rate is 627.

Reviewer: ```suggestion
The speed for the whole system creating 128 KiB files is 498 files/s, while the file reading speed is 627 files/s.
```

#### More References

> Here are some profiles available for simple local evaluation of file system performance. The specific test set size and number of concurrencies can be adjusted to suit for the practice.

Reviewer: ```suggestion
Here are some profiles available for simple local evaluation of file system performance. The specific test set size and number of concurrencies can be adjusted according to the actual situation.
```

##### Sequential reading and writing of large files

> All files are 1GiB in size, where `fwd1` is a large file for sequential writing, and `fwd2` is a large file for sequential reading.

Reviewer: ```suggestion
The file size is 1 GiB, where `fwd1` is a large file for sequential writing, and `fwd2` is a large file for sequential reading.
```

##### Random reading and writing of small files

> All files are 128KiB in size, where `fwd1` is a random small file for writing, `fwd2` is a random small file for reading, and `fwd3` is a random small file for mixed reading/writing (ratio read/write = 7:3).

Reviewer: ```suggestion
The file size is 128 KiB, where `fwd1` is a small file for random writing, `fwd2` is a small file for random reading, and `fwd3` is a small file for mixed reading/writing (read/write = 7:3).
```
